main_test.go

```diff
@@ -512,10 +515,71 @@ var testCases = []TestCase{
 	},
 }
 
+type FakeS3Client struct {
+	Get mockGetObjectAPI
+	Put mockPutObjectAPI
+}
+
+type mockGetObjectAPI func(ctx context.Context, params *s3.GetObjectInput) (*s3.GetObjectOutput, error)
+type mockPutObjectAPI func(ctx context.Context, params *s3.PutObjectInput) (*s3.PutObjectOutput, error)
+
+func (fc *FakeS3Client) GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	return fc.Get(ctx, params)
+}
+
+func (fc *FakeS3Client) PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error) {
+	return fc.Put(ctx, params)
+}
+
+func (fc *FakeS3Client) New(bucket, key string) *FakeS3Client {
+	return &FakeS3Client{
+		Get: mockGetObjectAPI(func(ctx context.Context, params *s3.GetObjectInput) (*s3.GetObjectOutput, error) {
+			if aws.StringValue(params.Bucket) != bucket {
+				return nil, fmt.Errorf("expected bucket %s; got %s", aws.StringValue(params.Bucket), bucket)
+			}
+			if aws.StringValue(params.Key) != key {
+				return nil, fmt.Errorf("expected key %s; got %s", aws.StringValue(params.Key), key)
+			}
+
+			f, err := os.Open(path.Join("/", key))
+			if err != nil {
+				return nil, fmt.Errorf("Error downloading s3 object: %s", err)
+			}
+			return &s3.GetObjectOutput{
+				Body: f,
+			}, nil
+		}),
+		Put: mockPutObjectAPI(func(ctx context.Context, params *s3.PutObjectInput) (*s3.PutObjectOutput, error) {
+			if aws.StringValue(params.Bucket) != bucket {
+				return nil, fmt.Errorf("expected bucket %s; got %s", aws.StringValue(params.Bucket), bucket)
+			}
+			if aws.StringValue(params.Key) != key {
+				return nil, fmt.Errorf("expected key %s; got %s", aws.StringValue(params.Key), key)
+			}
+
+			filename := path.Join("/", key)
+			body := params.Body
+			contents, err := io.ReadAll(body)
```

Comment:
fwiw you could use [`io.Copy`](https://pkg.go.dev/io#Copy) for less code and no unnecessary buffer. It even tells you the number of bytes. Not important though, up to you

Reply:
Yes, less code an no unnecessary buffer. Exactly what I need. Done.